File: 67-add-binary/add-binary.java
class Solution {
    public String addBinary(String a, String b) {
        StringBuilder result = new StringBuilder("");
        int carry = 0, i = a.length() - 1, j = b.length() - 1;

        while (i >= 0 || j >= 0 || carry == 1) {
            if (i >= 0)
                carry += a.charAt(i--) - '0';
            if (j >= 0) 
                carry += b.charAt(j--) - '0';

            result.append(carry % 2);
            carry = carry / 2;
        }
        return result.reverse().toString();
    }
}

File: 67-add-binary/add-binary.cpp
class Solution {
public:
    string addBinary(string a, string b) {
        string str;
        int m = a.length() - 1;
        int n = b.length() - 1;
        int carry = 0;

        while (m >= 0 || n >= 0) {
            int sum = carry;
            if (m >= 0)
                sum += a[m--] - '0';
            if (n >= 0)
                sum += b[n--] - '0';

            carry = sum > 1 ? 1 : 0;
            str += to_string(sum % 2);
        }
        if (carry)
            str += to_string(carry);
        reverse(str.begin(), str.end());
        return str;
    }
};